data/src/isupport.rs

```diff
@@ -630,9 +649,43 @@ pub struct CommandTargetLimit {
 }
 
 #[derive(Clone, Debug)]
+pub enum MessageReference {
+    Timestamp(DateTime<Utc>, String),
```

Comment:
What's the 2 field / string value of the `Timestamp` variant here? I don't see that documented anywhere here:

https://ircv3.net/specs/extensions/chathistory

This seems to be our own internal representation of a message reference _value_. I don't think this should live in `isupport`, but probably `message::Reference`

Reply:
I think this warrants discussion, particularly since I'm not 100% convinced the way I've implemented this is what we want long term.  Difficulties arise with `Timestamp` based `chathistory` since the `DateTime<Utc>` is not unique identifier.  All messages within the same second will have the same timestamp, and while that doesn't happen all the time it happens enough (especially since we're getting `JOIN`/`QUIT`/`PART`/etc messages as part of our requests).  Also, because of that lumping together, we need to request messages with a server time before the last message we saw, otherwise we might miss messages that happened in the same interval but the client didn't receive.  That means we will basically have at least one duplicate message per chathistory request when using `Timestamp`.  In order to perform deduplication and proper stitching together of message histories we need to be able to distinguish between messages with the same timestamp, which is where the `String` comes in.

Originally I had the `String` be the `text` of the message, but that has obvious issues when the message text depends on the current state (i.e. for `JOIN`/`QUIT`/etc messages where the user in `text` will be represented by what we currently know about them).  So, I switched the second bit of information to be a `client_id`, which is a seahash of the `Encoded` that comes in.  I don't love this really, since changing the structure of `Encoded` will cause ids to shift, but it works better than using `text`.

Client ids are distinguished from server ids by starting with `:`, which is an invalid character in a message tag (which is how server tags are delivered).

Reply:
I *swear* I was seeing second rounded server times, but I must have screwed up a debug print somewhere because I see proper millisecond times looking at it now.

Reply:
Pulling out the `client_id` business for now.  Will push that along with other changes to match our discussion tonight.

Reply:
@tarkah I figured out why I thought I was seeing server times rounded to the nearest second;  Halloy's public soju bouncer is returning message server times rounded to the nearest second.  The instance of soju I fired up for personal use was built from commit `6d0f7330e37360d791266bc31d026c4f2cfaad96` (the at the time latest commit on the master branch), and it's returning millisecond server times with messages.  Maybe the public bouncer is running an older version?